HACKING.jst.md

```diff
@@ -0,0 +1,100 @@
+  
+* `parser-standard` is meant to be very similar to the parser used in the
+  OCaml compiler. Its role in `ocamlformat` is (only) to provide a safety-check:
+  we want to make sure that the formatting does not change the AST as parsed by
+  the compiler. So after formatting, the parsed AST is checked against the
+  original parsed AST to make sure they are the same (modulo some normalization,
+  as written in `lib/Normalize_std_ast.ml`). Key point: this parser and its
+  representation are *not* pretty-printed and never interact with the
+  pretty-printer. The *only* reason this is here is to mimic the behavior of the
+  OCaml compiler. Accordingly, it should be as close to the OCaml compiler's
+  parser as possible.
+  
+* `parser-extended` uses an extended parsetree, capable of storing extra
+  information useful in preserving the structure of the user's code. It is this
+  extended parsetree that gets pretty-printed. The parser here forms part of the
+  core implementation of `ocamlformat`.
+  
+* `parser-recovery` is used for partial parsing, so that a bogus input source
+  file does not get mangled. It was an experiment that has been discontinued by
+  upstream and is not used with Jane Street. It uses the same parsetree as
+  `parser-extended`. A patchfile tracking the changes between `parser-extended`
+  and `parser-recovery` is generated. Just accept any changes that accrue there.
+  
+The directory `vendor/ocaml-common` contains files that are shared between
+`parser-standard` and `parser-extended`, like `location.ml`.
+  
+Testing is done in the `test` directory. TODO: Write more about the testing
+setup.
+
+Design considerations
+---------------------
+
+Because the value of `parser-standard` is entirely in its ability to mimic the
+compiler's parser, we want to keep this parser as close as possible to the
+compiler's. We will want to copy over any changes made to the compiler's parser
+into this version of `ocamlformat`'s parser.
+
+On the other hand, the `parser-extended` can go off in its own direction: its
+parsetree should be designed to make pretty-printing easy. In addition, we want
+to make sure that incorporating upstream changes is easy. We thus feel free to
+edit the parsetree in `parser-extended`, but we do so by adding new
+constructors, not modifying existing ones. In addition, new code should be
+marked off with comments like `(* Jane Street extension *)` and `(* End Jane
+Street extension *)`. Because of the ability to extend the parsetree here, we do
+*not* use jane-syntax in `parser-extended`.
+
+How to update `ocamlformat`
+---------------------------
+
+The base branch to work from is called `jane`. Create a branch off of `jane`.
+
+1. Take the patch you wish to support (i.e. some PR in `flambda-backend`).
+   Apply any changes to the `ocaml/parsing` directory to the files in
+   `vendor/parser-standard`. Remember: this "standard" parser should be as
+   close as possible to the compiler's.
```

Comment:
I think that one of the reasons that updating ocamlformat can be unpleasant, and which is particularly a reason that updating for the new jane syntax became very hard, is that people usually take the view of updating ocamlformat "for their PR/feature".  Perhaps we should instead take the view of incorporating all compiler changes since we last shipped an ocamlformat update.

That way, changes in the parser that don't actually need an ocamlformat update on their own still eventually end up being reflected in `parser-standard` (and perhaps in `parser-extended`, to the extent useful).  This keeps the standard parser close to the compiler's, and also reduces the (reasonably common) problem that you are trying to incorporate the changes from your PR, but the parser has changed in other ways in the meantime and figuring out what to take is a bit of a mess.

I'm not sure exactly how we'd organize that, though.  Perhaps we could record the `flambda-backend` commit hash that we believe ocamlformat to be "currently up to date with" somewhere, and the instructions would be to look at the `flambda-backend` diff since then, rather than just the diff of your PR?

Reply:
We are not the creators, owners, or maintainers of ocamlformat. We can't just move it, and maintaining a fork in a repo that does not even share history with the primary source does not seem great to me.

If we want to share code with the compiler, then let's have our patched compiler expose the necessary functionality in some user-facing library, and patch our import of ocamlformat to refer to that. That will be a minimal change to ocamlformat without having to duplicate lots of code or transplant a whole project.

Reply:
When I discussed this with Leo, he suggested a git subtree might be lighter weight than a git submodule.  That's not my experience with subtree, but it seems worth noting that some people like it.  Resolving this.